### PR TITLE
Add Semaphore locking support

### PR DESCRIPTION
**Pull request by athompson-r7**
This add support for obtaining a Semaphore lock[1], so that multiple nodes in a cluster can be preforming the same task at the same time.
- [x] Update README
- [x] Support locking on static limit
- [x] ~~Support locking on arbitrary dimensions, i.e. % of AWS Availability Zone~~ Instead rely on client code to figure this out
- [x] ~~Watch an existing lock for updates~~ Not really necessary(see example), proper support(#12)
- [x] Support releasing a lock after task is finished

[1] https://www.consul.io/docs/guides/semaphore.html

### CODE REVIEW

File: lib/daemon_runner/semaphore.rb [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    # Format the list of holders for the lock file
+    def holders
+      holders = {}
+      members = active_members
+      members << session.id if members.length < limit
+      members.map { |m| holders[m] = true }
+      holders
+    end
+
+    # Generate JSON formatted lockfile content, only if he number of contenders
```
Comment (fmitchell-r7) [inline lib/daemon_runner/semaphore.rb]: Spelling? "the" instead of "he"

File: lib/daemon_runner/semaphore.rb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      # Acquire a lock with the current session
+      #
+      # @param limit [Integer] The number of nodes that can request the lock
+      # @return [Boolean] `true` if the lock was acquired
+      #
+      def lock(limit = 3)
+        raise RuntimeError 'Must call start first' if @lock.nil?
+        @lock.limit = limit
+        @lock.contender_key
+        @lock.semaphore_state
```
Comment (fmitchell-r7) [inline lib/daemon_runner/semaphore.rb]: We need to check that the "Limit" in the semaphore state matches the `limit` provided as an argument. If they're not the same, we shouldn't try to write the lock.

Comment (athompson-r7) [inline lib/daemon_runner/semaphore.rb]: Addressed in b460287.

**Comment by athompson-r7** [comment]
The release lock code is blocked by a fix in diplomat[1], until it gets released we can point to the ref.

[1] WeAreFarmGeek/diplomat#105

**Comment by athompson-r7** [comment]
Fixes #12

**Pull request merged**